### ISSUE TITLE
Flaky VM crash in dart analyzer package_build_test.dart

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#36011]
This is breaking the CQ bots

```
/========================================================================================\
| pkg/analyzer/test/src/workspace/package_build_test failed again (Crash, expected Pass) |
\========================================================================================/

--- Command "vm" (took 20.000860s):
DART_CONFIGURATION=DebugX64 out/DebugX64/dart-sdk/bin/dart --no-background-compilation --enable_asserts --ignore-unrecognized-flags --packages=/b/s/w/ir/cache/builder/sdk/.packages /b/s/w/ir/cache/builder/sdk/pkg/analyzer/test/src/workspace/package_build_test.dart

exit code:
-6

stderr:
===== CRASH =====
version=2.2.0-edge.e4d1b286980ceace1f52d8ed3accb3b208db7c60 (Fri Feb 22 11:44:45 2019 +0000) on "linux_x64"
si_signo=Trace/breakpoint trap(5), si_code=128, si_addr=(nil)
thread=2760, isolate=kernel-service(0x556a68113b00)
  [0x00007fb0ea9d7713] Unknown symbol
  [0x00007fb0ea9d7713] Unknown symbol
  [0x00007fb0ea5de477] Unknown symbol
  [0x00007fb0ea67de18] Unknown symbol
  [0x00007fb0ea6f60ef] Unknown symbol
  [0x00007fb0e84dc88f] Unknown symbol
  [0x00007fb0ea66b060] Unknown symbol
  [0x00007fb0ea9adae4] Unknown symbol
  [0x00007fb0ea5e9181] Unknown symbol
  [0x00007fb0d9f65ec4] Unknown symbol
  [0x00007fb0e84fbd96] Unknown symbol
  [0x00007fb0db5f4e3b] Unknown symbol
  [0x00007fb0ea7f7e8b] Unknown symbol
  [0x00007fb0ea5eaee2] Unknown symbol
  [0x00007fb0ea62434b] Unknown symbol
  [0x00007fb0ea95791a] Unknown symbol
  [0x00007fb0ea67fe70] Unknown symbol
  [0x00007fb0e06c603c] Unknown symbol
  [0x00007fb0ea8ba048] Unknown symbol
  [0x00007fb0e84fde19] Unknown symbol
  [0x00007fb0ea8badf2] Unknown symbol
  [0x00007fb0ea8bb261] Unknown symbol
  [0x00007fb0ea8bbeb2] Unknown symbol
  [0x00007fb0ea955261] Unknown symbol
  [0x00007fb0ea8b7428] Unknown symbol
  [0x00007fb0ea8b6dfd] Unknown symbol
  [0x00007fb0ea952678] Unknown symbol
  [0x00007fb0ea95887b] Unknown symbol
  [0x00007fb0ea958a05] Unknown symbol
  [0x00007fb0ea6209fd] Unknown symbol
  [0x00007fb0ea6216e9] Unknown symbol
  [0x00007fb0ea61d6cd] Unknown symbol
  [0x00007fb0ea61d900] Unknown symbol
  [0x00007fb0ea5fd6dc] Unknown symbol
  [0x00007fb0eab8ef53] Unknown symbol
  [0x00007fb0ed4c153a] Unknown symbol
  [0x0000556a66755833] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
  [0x0000556a66758a61] dart::DartLibraryCalls::HandleMessage(dart::Object const&, dart::Instance const&)
  [0x0000556a667993ce] dart::IsolateMessageHandler::HandleMessage(dart::Message*)
  [0x0000556a667d8d88] dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
  [0x0000556a667d9af6] dart::MessageHandler::TaskCallback()
  [0x0000556a6696d17c] dart::ThreadPool::Worker::Loop()
  [0x0000556a6696cc5b] dart::ThreadPool::Worker::Main(unsigned long)
  [0x0000556a668b1795] Unknown symbol
-- End of DumpStackTrace

--- Re-run this test:
python tools/test.py -n unittest-asserts-debug-linux pkg/analyzer/test/src/workspace/package_build_test
```

**Comment by MichaelRFairhurst** [comment]
Could this be related as well?

```

/=========================================================================================\
| pkg/analyzer/test/src/summary/resynthesize_ast_test failed again (Crash, expected Pass) |
\=========================================================================================/

--- Command "vm" (took 05.000407s):
DART_CONFIGURATION=DebugX64 out/DebugX64/dart-sdk/bin/dart --no-background-compilation --enable_asserts --ignore-unrecognized-flags --packages=/b/s/w/ir/cache/builder/sdk/.packages /b/s/w/ir/cache/builder/sdk/pkg/analyzer/test/src/summary/resynthesize_ast_test.dart

exit code:
-6

stderr:
===== CRASH =====
version=2.2.0-edge.435a52eafa7e387f7296f6c181e392bf7d6a4dab (Wed Feb 20 23:15:47 2019 +0000) on "linux_x64"
si_signo=Trace/breakpoint trap(5), si_code=128, si_addr=(nil)
thread=19564, isolate=kernel-service(0x559f02ec5200)
  [0x00007faec7204e8a] Unknown symbol
  [0x00007faec7204e8a] Unknown symbol
  [0x00007faec6e97b23] Unknown symbol
  [0x00007faec720ca02] Unknown symbol
  [0x00007faec720dad6] Unknown symbol
  [0x00007faec72108ea] Unknown symbol
  [0x00007faec7069c1f] Unknown symbol
  [0x00007faec6f035c2] Unknown symbol
  [0x00007faec6f8c66b] Unknown symbol
  [0x00007faec6f6b93a] Unknown symbol
  [0x00007faec6f6c150] Unknown symbol
  [0x00007faec70bc420] Unknown symbol
  [0x00007faec70bc705] Unknown symbol
  [0x00007faec70bfc68] Unknown symbol
  [0x00007faec70bf330] Unknown symbol
  [0x00007faec70c0a12] Unknown symbol
  [0x00007faec70c0e81] Unknown symbol
  [0x00007faec726da92] Unknown symbol
  [0x00007faec6f79901] Unknown symbol
  [0x00007faec6f79fa8] Unknown symbol
  [0x00007faec6f7967d] Unknown symbol
  [0x00007faec6f88d38] Unknown symbol
  [0x00007faec6f8447b] Unknown symbol
  [0x00007faec6f84605] Unknown symbol
  [0x00007faec6f849fd] Unknown symbol
  [0x00007faec6f85da9] Unknown symbol
  [0x00007faec6f8696d] Unknown symbol
  [0x00007faec6f86ba0] Unknown symbol
  [0x00007faec6e0f11c] Unknown symbol
  [0x00007faec740b693] Unknown symbol
  [0x00007faec8c8153a] Unknown symbol
  [0x0000559f00505a13] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
  [0x0000559f00508c41] dart::DartLibraryCalls::HandleMessage(dart::Object const&, dart::Instance const&)
  [0x0000559f005495ae] dart::IsolateMessageHandler::HandleMessage(dart::Message*)
  [0x0000559f00588e28] dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
  [0x0000559f00589b96] dart::MessageHandler::TaskCallback()
  [0x0000559f0071cb6c] dart::ThreadPool::Worker::Loop()
  [0x0000559f0071c64b] dart::ThreadPool::Worker::Main(unsigned long)
  [0x0000559f006612c5] Unknown symbol
-- End of DumpStackTrace

--- Re-run this test:
python tools/test.py -n unittest-asserts-debug-linux pkg/analyzer/test/src/summary/resynthesize_ast_test
```

**Comment by stereotype441** [comment]
To clarify, this is not currently failing (https://ci.chromium.org/p/dart/builders/luci.dart.ci.sandbox/pkg-linux-debug/623).  But it was failing for builds 619-622, and we're not aware of any changes we made that would have provoked (or suppressed) the failure.  So it seems likely that we've triggered some flakiness in the VM that's still present.

**Comment by MichaelRFairhurst** [comment]
This also broke builds 602 & 603, if that stack trace is correct. We'll have to see if this is a continual problem or not

**Comment by mkustermann** [comment]
Does this reproduce consistently ? Do we have coredumps of this anywhere?

**Comment by a-siva** [comment]
Do we have coredump files from the crashes, without that or instructions on how to reliably reproduce the problem there is not much we can do with the stacktraces above.

**Comment by mkustermann** [comment]
As can be seen from the crash report above it hits a trap instruction (it's a Debug VM). I managed to get a dump.

This happens in a place where we store the context into a closure and we marked the store as not needing a write-barrier because it's preceded by a closure allocation. If that closure allocation happens on the slow path it could return an old object without having the remembered bit set, which hits the trap in DEBUG mode.

In release mode this could cause arbitrary VM crashes.

I'm working on a fix for it.
